### ISSUE TITLE
Conditional generation for forecasting translation task

### ISSUE THREAD

**Issue by JulianJB** [Y-debug-sys/Diffusion-TS#68]
I'm running a forecasting experiment with conditional generation for a data translation task using the sample `Tutorial_2.ipynb` notebook as a base, but I have a few questions regarding the Diffusion-TS implementation and how I can apply it to my specific use case.

Instead of using a train-test split as with the Solar dataset example, I’m trying to forecast an entire time series from a different but presumably correlated one used as the conditional target $y$ for the generation. Following the example, this is how I’m preparing my datasets:

`train_dataset = CustomDataset(data=train_scaled, regular=True, pred_length=24)`

`dataloader = DataLoader(dataset=train_dataset, batch_size=118, shuffle=False, num_workers=0, worker_init_fn=seed_worker, generator=g, drop_last=False, pin_memory=True, sampler=None)`

`test_dataset = CustomDataset(data=test_scaled, regular=False, pred_length=24)`

`test_dataloader = DataLoader(dataset=test_dataset, batch_size=118, shuffle=False, num_workers=0, worker_init_fn=seed_worker, generator=g, pin_memory=True, sampler=None)`

First, I’m trying to track the entire diffusion process (forward and reverse passes), but I noticed that when I run `trainer.train()`, the function called for calculating the loss value in the `solver.py` file is `loss.backward()`, which I assume is related to the denoising process. Before that, the loss is updated in the following line `loss = self.model(data, target=data)`, but is this where the model gets updated by adding Gaussian noise to the sample? I’m trying to locate the part where this function gets called during the training process (for the forward pass), which I believe has to be one of the functions in the `gaussian_diffusion.py` file (possibly`predict_noise_from_start(self, x_t, t, x0)`).

After that, I’m doing the conditional time series forecasting via the `trainer.restore()` function using the correlated time series as the test set. I notice that the conditional sampling method uses the `fast_sample_infill()` to get the samples (predictions), reals (original values taken from the test dataloader arrays), and masks (the booleans created from the test dataset) with the function `predict_noise_from_start(self, x_t, t, x0)` being called by the `fast_sample_infill()` and not the `predict_start_from_noise(self, x_t, t, noise)` function, which I think it should correspond to the reverse process. In the end, I can finally calculate the $MSE$  for the conditional generation via `mse = mean_squared_error(sample[~mask], real_test[~mask])` as in your example, but I’m not entirely sure if this masking approach is what I need for my use case.

My question here is, considering that the conditional sampling via the restore function is happening after the training of the model, is the test dataset still used as the conditional target $y$ for the generation? In other words, whenever the sampling is performed, either unconditionally via `trainer.sample()` or conditionally via `trainer.restore()`, does this process corresponds to the reverse process of the diffusion model where the only difference is the addition of a conditional variable $y$ as the target for the generation? Is this correct?

As I mentioned in the beginning, what I’m trying to do is to track the workflow of the diffusion process for both passes (particularly for the reverse pass) since I believe my particular use case involves some kind of guided diffusion where the target time series observed is different from the one used for the forward pass and is guiding the reconstruction by adding new information to the denoising process in a similar way as the discriminator component works in a traditional GAN model.

I’ve digging through some of the threads here on GitHub raised by other users found the following ones particularly interesting, so I’m soft-linking them to this post in the hopes of finding some connection to questions previously reported: [[#53](https://github.com/Y-debug-sys/Diffusion-TS/issues/53)] [[#40](https://github.com/Y-debug-sys/Diffusion-TS/issues/40)] [[#66](https://github.com/Y-debug-sys/Diffusion-TS/issues/66)] [[#41](https://github.com/Y-debug-sys/Diffusion-TS/issues/41)] [[#60](https://github.com/Y-debug-sys/Diffusion-TS/issues/60)] [[#30](https://github.com/Y-debug-sys/Diffusion-TS/issues/30)]

Thanks in advance for your help and for open-sourcing your work, it’s a valuable contribution to time series research.

**Comment by Y-debug-sys** [comment]
Hi, JulianJB. In response to your problem, my answer is as follows:

> Instead of using a train-test split as with the Solar dataset example, I’m trying to forecast an entire time series from a different but presumably correlated one used as the conditional target $y$ for the generation.

**A:** Does `"different but presumably correlated"` mean something similar? If so, please refer to the previous [issue](https://github.com/Y-debug-sys/Diffusion-TS/issues/62). If not, the approach used in the codebase must explicitly specify the relationship between these two sequences (then use a loss to force the relationship). Additionally, I noticed that you aim to generate an entire series, but note that our code currently does not support generating extremely long sequences owing to the long sampling time.

> First, I’m trying to track the entire diffusion process (forward and reverse passes), but I noticed that when I run `trainer.train()`, the function called for calculating the loss value in the `solver.py` file is `loss.backward()`, which I assume is related to the denoising process. Before that, the loss is updated in the following line `loss = self.model(data, target=data)`, but is this where the model gets updated by adding Gaussian noise to the sample? I’m trying to locate the part where this function gets called during the training process (for the forward pass), which I believe has to be one of the functions in the `gaussian_diffusion.py` file (possibly`predict_noise_from_start(self, x_t, t, x0)`).

**A:** Yes, your understand is OK.

> My question here is, considering that the conditional sampling via the restore function is happening after the training of the model, is the test dataset still used as the conditional target $y$ for the generation? In other words, whenever the sampling is performed, either unconditionally via `trainer.sample()` or conditionally via `trainer.restore()`, does this process corresponds to the reverse process of the diffusion model where the only difference is the addition of a conditional variable $y$ as the target for the generation? Is this correct?

**A:** Yes, our method only requires an unconditional diffusion model to handle both conditional and unconditional generation. The difference between two generation tasks is that, in conditional generation, additional gradient-based conditioning imposed by a relationship *equation* is applied during steps of the reverse process.

> As I mentioned in the beginning, what I’m trying to do is to track the workflow of the diffusion process for both passes (particularly for the reverse pass) since I believe my particular use case involves some kind of guided diffusion where the target time series observed is different from the one used for the forward pass and is guiding the reconstruction by adding new information to the denoising process in a similar way as the discriminator component works in a traditional GAN model.

**A:** I believe the discriminator component you proposed is feasible, by referring to the idea of [Classifier Guidance](https://arxiv.org/abs/2105.05233). Suppose there is a pre-trained discriminator network $D(x_{t.c}, x_{t.g}, t)$ used to determine whether the two sequences satisfy the relationship you described. Then, during each diffusion step $t$ in the reverse process, you can control the model's output to make $D$ approach a value of 1 (using gradient-based guidance as in the code), for achieving your goal.

**Comment by JulianJB** [comment]
Thank you for your comprehensive reply and for the resources linked, I will carefully look into them. And yes, it would mean that both time series are similar and also, to some extent, correlated.

I have a follow-up question regarding the masking process. I see that the masking is applied to the data while initiliazing the CustomDataset object in the following two lines inside the `__init__` function:

`self.mask = np.ones_like(data)`

`self.mask[:, -pred_length:, :] = 0.`

And then, when we set `regular=False` as a parameter for the test set and call the `trainer.restore()` function, the array of masked values is filled via `mask = self.mask[ind, :, :]` inside the `__getitem__` function, which is used as the partial mask`t_m` for the `fast_sample_infill` function.

As I understand from your paper, the masking strategy is used to control the length of the sequences established by the missing ratio of the dataset.

Now that `t_m` is the partial mask for the test set, the sample prediction for forecasting would be the result of the product of the real unmasked values of the test set denoted by `x` and the masked values provided by the partial mask of the test set denoted by `t_m`. So the target for the conditional sample is `x*t_m`, which then goes through some additional functions from the `gaussian_diffusion.py` trainer class, including the `langevin_fn` function and the matching `img[partial_mask] = target_t[partial_mask]`.

So, for example, if I’m passing the split of my test set with shape `(1180, 24, 3)`, the array of the masked values would be converted to `0.` and then to boolean `False` afterwards with the same shape. Does this mean that the method is doing forecasting for the whole sequence passed? I notice that if I print the shape of the mask array, it returns `(24, 3)`, so it’s masking the values via the `fast_sample_infill` function until it has the same size as the real samples of the test set, thus `84960`. So this means that I can technically plot any segment of the time series returned from the conditional sampling against the original time series used as the test set since all the values of the test set were observed at some point during the forecasting task. Is this correct?

Can you elaborate a bit more on how the masking strategy works for the forecasting task?

Thank you again for your help in providing a clearer insight into your work.

**Comment by Y-debug-sys** [comment]
> So this means that I can technically plot any segment of the time series returned from the conditional sampling against the original time series used as the test set since all the values of the test set were observed at some point during the forecasting task. Is this correct? Can you elaborate a bit more on how the masking strategy works for the forecasting task?

Yes, you're understanding it correctly! For the forecasting task, we simply mask the part to be predicted (or add empty values after the known sequence), and then guide the diffusion model to reconstruct the known part (through masked MSE or MAE) to generate the masked part.
